walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,70 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.Collections;
+import java.util.ArrayList;
+import java.util.List;
 
 public final class FindMeetingQuery {
+  
+  public final Collection<TimeRange> slots = new ArrayList<>();
+  public final List<TimeRange> bookedTimes = new ArrayList<>();
+
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+
+    if(request.getDuration() >= TimeRange.WHOLE_DAY.duration()){
+        return slots;
+    }else if(events.isEmpty()){
+        slots.add(TimeRange.WHOLE_DAY);
+    }else {
+        for(Event event: events){
+            for(String attendee: request.getAttendees()){
+                if(event.getAttendees().contains(attendee)){
+                    bookedTimes.add(event.getWhen());
+                    break;
+                }
+            }
+        }
+
+        if(bookedTimes.size() == 0){
+            slots.add(TimeRange.WHOLE_DAY);
+            return slots;
+        }
+
+        //Sort the booked times by start of day
+        Collections.sort(bookedTimes, TimeRange.ORDER_BY_START);
+
+        //Merge overlapping or nested timeslots into a single chunk of time
+        for (int i = 0; i < bookedTimes.size() - 1; i++){
+            if(bookedTimes.get(i).overlaps(bookedTimes.get(i+1)) || bookedTimes.get(i).contains(bookedTimes.get(i+1))){
+                if(bookedTimes.get(i).end() > bookedTimes.get(i+1).end()){
+                    bookedTimes.set(i, TimeRange.fromStartEnd(bookedTimes.get(i).start(), bookedTimes.get(i).end(), false));
+                }else{
+                    bookedTimes.set(i, TimeRange.fromStartEnd(bookedTimes.get(i).start(), bookedTimes.get(i+1).end(), false));
+                }
+                bookedTimes.remove(i+1);
+            }
+        }
+
+        //Get everthing in between the booked time slots as eligible slots for the request
+        long duration = request.getDuration();
+        
+        if((bookedTimes.get(0).start() - TimeRange.START_OF_DAY) >= duration){
+            slots.add(TimeRange.fromStartEnd(TimeRange.START_OF_DAY, bookedTimes.get(0).start(), false));
+        }
+
+        for(int i = 0; i < bookedTimes.size() - 1; i++){
+            if((bookedTimes.get(i+1).start() - bookedTimes.get(i).end()) >= duration){
```

Comment:
Ah I think this check allowed you to correct your logical error in the above for loop. If the problem had just asked for all free slots (of any duration), then you wouldn't have this check and then erroneous slots would have been added to the final result.

Reply:
This just allows my function to only filter the slots that are at least the size of duration

Reply:
Yes but if the problem didn't ask for a specific duration and instead asked only to return all free slots, you wouldn't have had this check because then all you would do is add the slots in between merged bookedTimes intervals. In that case the initial error in the for loop would have caused incorrect slots being added.